service/qos/service_level_controller.cc

```diff
@@ -219,16 +220,22 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
 
 future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {
```

Comment:
I think this should be refactored differently, we can enhance the subscriber api by adding "before" and "after" version. That way it will be more verbose and you will not need to supply one more key when subscribing.
for example, on_service_level_add, will become two functions on_before_service_level_add and on_after_service_level_add

Reply:
Agree.

There's the question of exception safety (preexisting). If one of the notifiers throws, how do we handle it?

Reply:
I guess we could add a cancel notification. "before" would be allowed to throw, but cancel and after not. "before" would be required to reserve all necessary resources.